langchain4j-vertex-ai-gemini/src/test/java/dev/langchain4j/model/vertexai/VertexAiGeminiChatModelIT.java

```diff
@@ -213,38 +221,85 @@ void should_accept_text_and_multiple_base64_images() {
-                TextContent.from("What do you see? Reply with one word per image.")
+            ImageContent.from(CAT_IMAGE_URL),
+            ImageContent.from("gs://langchain4j-test/dog.jpg"),
+            ImageContent.from(Base64.getEncoder().encodeToString(readBytes(DICE_IMAGE_URL)), "image/png"),
+            TextContent.from("What do you see? Reply with one word per image.")
         );
 
         // when
         Response<AiMessage> response = visionModel.generate(userMessage);
 
         // then
         assertThat(response.content().text())
-                .containsIgnoringCase("cat")
-                .containsIgnoringCase("dog")
-                .containsIgnoringCase("dice");
+            .containsIgnoringCase("cat")
+            .containsIgnoringCase("dog")
+            .containsIgnoringCase("dice");
+    }
+
+    @Test
+    void should_accept_tools_for_function_calling() {
+
+        // given
+        ChatLanguageModel model = VertexAiGeminiChatModel.builder()
+            .project(System.getenv("GCP_PROJECT_ID"))
+            .location(System.getenv("GCP_LOCATION"))
+            .modelName("gemini-pro")
+            .build();
+
+        ToolSpecification weatherToolSpec = ToolSpecification.builder()
+            .name("getWeatherForecast")
+            .description("Get the weather forecast for a location")
+            .addParameter("location", JsonSchemaProperty.STRING,
+                JsonSchemaProperty.description("the location to get the weather forecast for"))
+            .build();
+
+        List<ChatMessage> allMessages = new ArrayList<>();
+
+        UserMessage weatherQuestion = UserMessage.from("What is the weather in Paris?");
+        System.out.println("Question: " + weatherQuestion.text());
+        allMessages.add(weatherQuestion);
+
+        // when
+        Response<AiMessage> messageResponse = model.generate(allMessages, weatherToolSpec);
+
+        // then
+        assertThat(messageResponse.content().hasToolExecutionRequests()).isTrue();
+        ToolExecutionRequest toolExecutionRequest = messageResponse.content().toolExecutionRequests().get(0);
+
+        assertThat(toolExecutionRequest.arguments()).contains("Paris");
+        assertThat(toolExecutionRequest.name()).isEqualTo("getWeatherForecast");
+
+        allMessages.add(messageResponse.content());
+
+        // when (feeding the function return value back)
+        ToolExecutionResultMessage toolExecResMsg = ToolExecutionResultMessage.from(toolExecutionRequest,
+            "{\"location\":\"Paris\",\"forecast\":\"sunny\", \"temperature\": 20}");
+        allMessages.add(toolExecResMsg);
+
+        Response<AiMessage> weatherResponse = model.generate(allMessages);
+
+        // then
+        System.out.println("Answer: " + weatherResponse.content().text());
+        assertThat(weatherResponse.content().text()).containsIgnoringCase("sunny");
     }
 }
```

Comment:
minor: would be awesome to extend tests in `AiServicesWithToolsIT` to be parameterized with the model (same way as in `AiServicesWithRagIT`) and add Gemini model to the OpenAI.